lib/algae/result.ex

```diff
@@ -0,0 +1,323 @@
+  end
+end
+
+defimpl TypeClass.Property.Generator, for: Ok do
+  def generate(_) do
+    [1, 1.1, "", []]
+    |> Enum.random()
+    |> TypeClass.Property.Generator.generate()
+    |> Ok.new()
+  end
+end
+
+##########
+# Setoid #
+##########
+
+definst Witchcraft.Setoid, for: Err do
+  def equivalent?(_, %Ok{}), do: false
+  def equivalent?(%Err{err: a}, %Err{err: b}), do: Witchcraft.Setoid.equivalent?(a, b)
+end
+
+definst Witchcraft.Setoid, for: Ok do
+  def equivalent?(_, %Err{}), do: false
+  def equivalent?(%Ok{ok: a}, %Ok{ok: b}), do: Witchcraft.Setoid.equivalent?(a, b)
+end
+
+#######
+# Ord #
+#######
+
+definst Witchcraft.Ord, for: Err do
+  custom_generator(_) do
+    1
+    |> TypeClass.Property.Generator.generate()
+    |> Err.new()
+  end
+
+  def compare(_, %Ok{}), do: :lesser
+  def compare(%Err{err: a}, %Err{err: b}), do: Witchcraft.Ord.compare(a, b)
+end
+
+definst Witchcraft.Ord, for: Ok do
+  custom_generator(_) do
+    1
+    |> TypeClass.Property.Generator.generate()
+    |> Ok.new()
+  end
+
+  def compare(_, %Err{}), do: :greater
+  def compare(%Ok{ok: a}, %Ok{ok: b}), do: Witchcraft.Ord.compare(a, b)
+end
+
+#############
+# Semigroup #
+#############
+
+definst Witchcraft.Semigroup, for: Err do
+  custom_generator(_) do
+    1
+    |> TypeClass.Property.Generator.generate()
+    |> Err.new()
+  end
+
+  def append(err, %Ok{}), do: err
+  def append(%Err{err: a}, %Err{err: b}), do: %Err{err: a <> b}
+end
+
+definst Witchcraft.Semigroup, for: Ok do
+  custom_generator(_) do
+    1
+    |> TypeClass.Property.Generator.generate()
+    |> Ok.new()
+  end
+
+  def append(_, err = %Err{}), do: err
+  def append(%Ok{ok: a}, %Ok{ok: b}), do: %Ok{ok: a <> b}
+end
+
+##########
+# Monoid #
+##########
+
+definst Witchcraft.Monoid, for: Err do
+  def empty(%Err{err: a}), do: %Ok{ok: Witchcraft.Monoid.empty(a)}
+end
+
+definst Witchcraft.Monoid, for: Ok do
+  def empty(%Ok{ok: a}), do: %Ok{ok: Witchcraft.Monoid.empty(a)}
+end
+
+###########
+# Functor #
+###########
+
+definst Witchcraft.Functor, for: Err do
+  def map(err, _), do: err
+end
+
+definst Witchcraft.Functor, for: Ok do
+  def map(%Ok{ok: data}, fun), do: data |> fun.() |> Ok.new()
+end
+
+############
+# Foldable #
+############
+
+definst Witchcraft.Foldable, for: Err do
+  def right_fold(_, seed, _), do: seed
+end
+
+definst Witchcraft.Foldable, for: Ok do
+  def right_fold(%Ok{ok: inner}, seed, fun), do: fun.(inner, seed)
+end
+
+###############
+# Traversable #
+###############
+
+definst Witchcraft.Traversable, for: Err do
+  @force_type_instance true
+
+  def traverse(err = %Err{err: value}, link) do
+    value
+    |> link.()
+    |> of(err)
+  end
+end
+
+definst Witchcraft.Traversable, for: Ok do
+  def traverse(%Ok{ok: value}, link), do: map(link.(value), &Ok.new/1)
+end
+
+#########
+# Apply #
+#########
+
+definst Witchcraft.Apply, for: Err do
+  def convey(err, _), do: err
+end
+
+definst Witchcraft.Apply, for: Ok do
+  def convey(_,    %Err{}), do: %Err{}
```

Comment:
Also, do you want this to drop the `any` in `%Err{}`?

Have you considered this?

```elixir
def convey(_,    err = %Err{}), do: err
```

This keeps looking a lot like a renamed `Either`. No? Halp?

Reply:
I will look into making `Err` an `Exception` today.
Regarding `convey`, you're right, that should be changed. Well spotted, thanks! 

It's totally a renamed `Either`, that's the first thing I said ages ago about this 😜 
I needed something with a bit more "context", otherwise I'd forget what left or right is 🤷‍♂️